### PR TITLE
Refactor to use the  useContext hook

### LINKED ISSUE #1569: Use Hooks for Context - useContext instead of Context Consumer
**Is your feature request related to a problem? Please describe.**
We're currently using the Context API in various places. We should move to using the Hooks implementation of it with `useContext`. Here is a description of the benefits of using the React Hooks implementation of Context https://frontarm.com/james-k-nelson/usecontext-react-hook/ 

https://github.com/bbc/simorgh/blob/ead571daf45100e1749fe3be96c2eef8deaec639/src/app/containers/ConsentBanner/Banner/index.amp.jsx#L35
https://github.com/bbc/simorgh/blob/4cbe0a30ff6b7976d415800138b3bdabb7fdb173/src/app/containers/Caption/index.jsx#L19
https://github.com/bbc/simorgh/blob/350761cbc024edd2ca206190b3b105827376aafc/src/app/containers/Brand/index.jsx#L6
https://github.com/bbc/simorgh/blob/34450a370a1c5df736622d7b0a470d996c6e8fc5/src/app/containers/Copyright/index.jsx#L9
https://github.com/bbc/simorgh/blob/ead571daf45100e1749fe3be96c2eef8deaec639/src/app/containers/ConsentBanner/Banner/index.canonical.jsx#L20
https://github.com/bbc/simorgh/blob/fdd997630c43819a36a1390d1666778be7d31929/src/app/containers/ErrorMain/index.jsx#L33
https://github.com/bbc/simorgh/blob/8cc14f8b7b138b6a7c9a841749828c8d028189c5/src/app/containers/Metadata/index.jsx#L30-L32

**Describe the solution you'd like**
To use `useContext(ServiceContext)`, `useContext(RequestContext)` instead of `<ServiceContext.Consumer>` `<RequestContext.Consumer>` in this codebase. 

**Describe alternatives you've considered**
We could leave the structure as it currently is, since it will continue to work in the same way. This work enhances readability for the usage of Contexts. 

**Testing notes**
[Tester to complete]

Dev insight: Snapshots would need to be updated. Other unit tests should continue to pass. Cypress tests should continue to pass. 

- [x] Initially labelled with ["Refinement needed"](https://github.com/bbc/simorgh/labels/Refinement%20Needed)

### PR DESCRIPTION
**Pull request by chris-hinds**
Resolves #1569 

**Overall change:**
Removes the need for wrapping containers in a ContextConsumer to access the various contexts objects by utilising the useContext hook.

Note: This PR removes enzyme tests from the headings container replacing them with snapshots to support the use of react hooks. 

The headings container was the only outlier that was causing problems when refactoring to use the useContext hook.

However I know that some pods are starting to switch to enzyme, looking to replace snapshots on containers with an assertion library. With this in mind I think the use of snapshots here right now is ok (considering most of our containers have used them/are still using them). Then when we agree as a team on the assertion library to use this can be refactored again to use the new testing library.

**Code changes:**

- Replaces context consumers with the useContext hook

---

- [X] I have assigned myself to this PR and the corresponding issues
- [X] Tests added for new features (Replaced enzyme with snapshots as enzyme does not yet support React Hooks)
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

File: src/app/containers/Headings/index.test.jsx [hunk 1]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
+    describe('should render headline', () => {
       const data = {
         blocks: blocksSingleFragment('Plain headline', []),
         type: 'headline',
       };
-      const renderedWrapper = render(HeadingsContainerWithContext(data));
 
-      expect(renderedWrapper.find('h1').html()).toBe('Plain headline');
+      shouldMatchSnapshot(
+        'should render correctly',
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx]:
```suggestion
        'should render h1 containing correct text',
```

File: src/app/containers/Headings/index.test.jsx [hunk 2]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
+    describe('should render subheadline', () => {
       const data = {
         blocks: blocksSingleFragment('Plain subheadline', []),
         type: 'subheadline',
       };
-      const renderedWrapper = render(HeadingsContainer(data));
 
-      expect(renderedWrapper.find('h2').html()).toBe('Plain subheadline');
+      shouldMatchSnapshot(
+        'should render correctly',
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx]:
```suggestion
        'should render h2 containing correct text',
```

File: src/app/containers/Headings/index.test.jsx [hunk 3]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
+    describe('should render headline with italic text', () => {
       const data = {
         blocks: blocksSingleFragment('All is italic', ['italic']),
         type: 'headline',
       };
-      const renderedWrapper = render(HeadingsContainer(data));
 
-      expect(renderedWrapper.find('h1').html()).toBe('<i>All is italic</i>');
+      shouldMatchSnapshot(
+        'should render correctly',
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx]:
```suggestion
        'should render h1 with <i> tag',
```

File: src/app/containers/Headings/index.test.jsx [hunk 4]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
+    describe('should render headline with bold text', () => {
       const data = {
         blocks: blocksSingleFragment('All is bold', ['bold']),
         type: 'headline',
       };
-      const renderedWrapper = render(HeadingsContainer(data));
 
-      expect(renderedWrapper.find('h1').html()).toBe('<b>All is bold</b>');
+      shouldMatchSnapshot(
+        'should render correctly',
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx]:
```suggestion
        'should render h1 with <b> tag',
```

File: src/app/containers/Headings/index.test.jsx [hunk 5]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
           'italic',
         ]),
         type: 'headline',
       };
-      const renderedWrapper = render(HeadingsContainer(data));
 
-      expect(renderedWrapper.find('h1').html()).toBe(
-        '<i><b>All is bold and italic</b></i>',
+      shouldMatchSnapshot(
+        'should render correctly',
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx]:
```suggestion
        'should render h1 with <i><b> tags',
```

File: src/app/containers/Headings/index.test.jsx [hunk 6]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
           'bold',
         ]),
         type: 'headline',
       };
-      const renderedWrapper = render(HeadingsContainer(data));
 
-      expect(renderedWrapper.find('h1').html()).toBe(
-        '<b><i>All is italic and bold</i></b>',
+      shouldMatchSnapshot(
+        'should render correctly',
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx]:
```suggestion
        'should render h1 with <b><i> tags',
```

File: src/app/containers/Headings/index.test.jsx [hunk 7]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
         const data = {
           blocks: textItalicFragmentPart('This is ', 'very', ' important'),
           type: 'headline',
         };
-        const renderedWrapper = render(HeadingsContainer(data));
 
-        expect(renderedWrapper.find('h1').html()).toBe(
-          'This is <i>very</i> important',
+        shouldMatchSnapshot(
+          'should render correctly',
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx:188]:
```suggestion
          'should render h1 with <i> around fragment',
```

File: src/app/containers/Headings/index.test.jsx [hunk 8]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
-        '<b><i>All is italic and bold</i></b>',
+      shouldMatchSnapshot(
+        'should render correctly',
+        <HeadingsContainer {...data} />,
       );
     });
 
     describe('with different attributes', () => {
-      it('should render headline with only middle of text italic', () => {
+      describe('should render headline with only middle of text italic', () => {
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx:181]:
```suggestion
      describe('headline with only middle of text italic', () => {
```

File: src/app/containers/Headings/index.test.jsx [hunk 9]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
-      expect(renderedWrapper.find('h1').html()).toBe(
-        '<i><b>All is bold and italic</b></i>',
+      shouldMatchSnapshot(
+        'should render correctly',
+        <HeadingsContainer {...data} />,
       );
     });
 
-    it('should render headline with italic & bold text', () => {
+    describe('should render headline with italic & bold text', () => {
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx:165]:
```suggestion
    describe('headline with italic & bold text', () => {
```

File: src/app/containers/Headings/index.test.jsx [hunk 10]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
 
-      expect(renderedWrapper.find('h1').html()).toBe('<b>All is bold</b>');
+      shouldMatchSnapshot(
+        'should render correctly',
+        <HeadingsContainer {...data} />,
+      );
     });
 
-    it('should render headline with bold & italic text', () => {
+    describe('should render headline with bold & italic text', () => {
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx:150]:
```suggestion
    describe('headline with bold & italic text', () => {
```

File: src/app/containers/Headings/index.test.jsx [hunk 11]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
   template('with subheadline data', 'This is a subheadline', 'subheadline');
 
   describe('with plain text', () => {
-    it('should render headline', () => {
+    describe('should render headline', () => {
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx]:
```suggestion
    describe('headline', () => {
```

File: src/app/containers/Headings/index.test.jsx [hunk 12]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
 
-      expect(renderedWrapper.find('h1').html()).toBe('Plain headline');
+      shouldMatchSnapshot(
+        'should render correctly',
+        HeadingsContainerWithContext(data),
+      );
     });
 
-    it('should render subheadline', () => {
+    describe('should render subheadline', () => {
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx]:
```suggestion
    describe('subheadline', () => {
```

File: src/app/containers/Headings/index.test.jsx [hunk 13]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
+      shouldMatchSnapshot(
+        'should render correctly',
+        <HeadingsContainer {...data} />,
+      );
     });
   });
 
   describe('with rich text', () => {
-    it('should render headline with italic text', () => {
+    describe('should render headline with italic text', () => {
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx]:
```suggestion
    describe('headline with italic text', () => {
```

File: src/app/containers/Headings/index.test.jsx [hunk 14]
```diff
@@ -98,88 +97,96 @@ describe('Headings', () => {
...
 
-      expect(renderedWrapper.find('h1').html()).toBe('<i>All is italic</i>');
+      shouldMatchSnapshot(
+        'should render correctly',
+        <HeadingsContainer {...data} />,
+      );
     });
 
-    it('should render headline with bold text', () => {
+    describe('should render headline with bold text', () => {
```
Comment (sareh) [inline src/app/containers/Headings/index.test.jsx]:
```suggestion
    describe('headline with bold text', () => {
```

File: src/app/containers/InlineLink/index.jsx [hunk 15]
```diff
@@ -33,16 +34,12 @@ const InlineLinkContainer = ({ locator, isExternal, blocks }) => {
...
-      {({ externalLinkText }) => (
-        <InlineLink href={locator}>
-          <Blocks blocks={blocks} componentsToRender={componentsToRender} />
-          {isExternal ? (
-            <VisuallyHiddenText>{externalLinkText}</VisuallyHiddenText>
-          ) : null}
-        </InlineLink>
-      )}
-    </ServiceContextConsumer>
+    <InlineLink href={locator}>
```
Comment (sareh) [inline src/app/containers/InlineLink/index.jsx:37]: It's great how using `useContext` makes the code much more readable. ✨

**Review by sareh — COMMENTED** [review]
Looks great. I've just added suggestions for the naming of unit tests, where we've changed the tests to be snapshots.

**Comment by chris-hinds** [comment]
Thanks for the suggestions @sareh all suggestions applied.

File: src/app/containers/Figure/index.jsx [hunk 16]
```diff
@@ -20,30 +20,29 @@ const FigureContainer = ({
...
-    </ImagePlaceholder>
-    {renderCaption(captionBlock)}
-  </Figure>
-);
+}) => {
+  const { platform } = useContext(RequestContext);
+  return (
+    <Figure>
+      <ImagePlaceholder ratio={ratio}>
+        {platform === 'amp' ? (
```
Comment (sareh) [inline src/app/containers/Figure/index.jsx:30]: Now that the context has been moved to the top of this function, we should check to see whether the whole `<Figure>` including caption, copyright & placeholder re-render when a platform context value is updated, or whether it is clever enough to only re-render the actual Image part of this container.

Comment (chris-hinds) [reply src/app/containers/Figure/index.jsx:30]: I would have thought the whole image would still re-render. Dp you know what currently happens @sareh 

Also the only platform change would be between an amp and a canonical page right? do what re-renders actually matter in this use case?

Comment (sareh) [reply src/app/containers/Figure/index.jsx:30]: You're correct about this usecase in that it's an unlikely user journey. 

Where we pull in translations via the Service Context, e.g. https://github.com/bbc/simorgh/pull/1656/files#diff-5192a5db9928faa2474d9f458c25d32cR36, we don't want to re-render the whole component every time a particular translation changes. We may need to use the useMemo hook to ensure a re-render only happens when relevant values are updated.

**Review by sareh — APPROVED** [review]
👍 Looks great! Thanks for updating this.

**Review by jamesdonoh — APPROVED** [review]

**Comment by PriyaKR** [comment]
LGTM

**Pull request merged**